### ISSUE TITLE
Preserve argument types when applying higher-order functions

### ISSUE THREAD

**Issue by sekwiatkowski** [microsoft/pyright#400]
**Is your feature request related to a problem? Please describe.**
Not entirely sure if this is a bug of works as intended. Let's say I want to define a higher-order function that composes two functions:

```Python
T = TypeVar('T')
U = TypeVar('U')
V = TypeVar('V')

def compose2(f: Callable[[T], U], g: Callable[[U], V]) -> Callable[[T], V]:
    def composition(x: T) -> V:
        return g(f(x))
    return composition

def add_one(x: int) -> int:
    return x + 1

# (p0: Any) -> int instead of (p0: int) -> int
add_two = compose2(add_one, add_one)
```

The following type is inferred for add_two: (p0: Any) -> int.

**Describe the solution you'd like**
Is there a way to write this such that the type of the argument of the first function is preserved? I would like the inferred type of add_two to be (p0: int) -> int.

**Comment by erictraut** [comment]
Thanks for the bug report. This is indeed a bug in pyright. It wasn't correctly handling contravariant type var matching in some cases. All callable input parameters are contravariant, which is why the bug affected your example. I've fixed the bug and added your example as a new test case.

**Comment by sekwiatkowski** [comment]
Thanks so much for changing this! You have no idea how much this means to me :-) For years I've been dreaming of using at least some of the basic principles of functional programming in Python with type checking.

I understand that Python is not exactly designed for that, but I will see how far I can get with PyRight.

**Comment by erictraut** [comment]
This is now implemented in version 1.1.9, which I just published.
